[WIP] Add CalicoNodeStatus resource

song-jiang opened the pull request:
## Description
<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note
<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
+
+	// BirdStatus represents the latest observed status of bird.
+	BirdStatus CalicoNodeBirdStatus `json:"birdStatus,omitempty"`
```
Does this belong in BGP status or in Agent status?

I suspect in Agent status, since BIRD is capable of doing more than just BGP (even though we largely use it only for BGP today).

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
```
Wondering if we should have calico/node create one of these as part of its startup logic, and then users can simply `kubectl edit` it in order to enable / configure it.

If we default to `0s` here, then we won't start reporting until asked for. WDYT?

song-jiang replied:
I'm not very keen on that. Because
- a large cluster may create loads of NodeStatus objects but actually users are not going to use them at all. This resource is still more or less for debugging purpose. 
- User may just want to focus on one node but get lost on hundreds of objects listed with a `list` api call. 
- It is easier to create a resource correctly than to patch/edit it. Also easier to document as well.

caseydavenport replied:
Agree more or less with the first two bullet points, but strongly disagree with ease of documentation. Easier to follow documentation was my main reason for suggesting this.

All of our documentation is going to have to be written in the form "If this object exists, then modify it. If it doesn't exist, create it".

Which is somewhat confusing

caseydavenport replied:
This needs to be a `*int` if we want to perform defaulting on it - otherwise the code will see it as `0` and assume the user disabled refresh.

neiljerram replied:
I think the documentation point depends on what we consider the mainline use case to be.  If it is that these resources are only used when there is a specific need to investigate or understand something, then the documentation could say "Create this object, and read it back to see node debugging information.  When you've resolved the issue concerned, delete the object again."

Given that this object's identity is NOT the node name, and that there is only one Interval field per object, I feel that's the right model.  For example, suppose there is an existing CalicoNodeStatus for node MYNODE and classes Grok and Frobnicate: you can still create a new CalicoNodeStatus (called e.g. "mynode-bgp36") for node MYNODE to monitor BIRD stuff.

neiljerram replied:
BTW - and I don't know if this is even possible - but did you consider a model where status is reported as a status subresource of the Calico Node object?

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	UpdateTimestamp metav1.Time `json:"updateTimestamp,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
```
Should this be a map[string]string so that we can add arbitrary metadata to it?

Remind me what sort of information we're planning to put here today?

song-jiang replied:
This is just an additional information about the status of last update. It could contain error message if the last update failed for any reason.

caseydavenport replied:
Kubernetes generally uses `Conditions` in status to report on things like this - we should follow that pattern instead of adding an opaque string. It's a more structured way of reporting messages intended for the end-user / a human.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+type CalicoNodeStatusStatus struct {
+	// UpdateTimestamp is a timestamp representing the server time when CalicoNodeStatus object
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	UpdateTimestamp metav1.Time `json:"updateTimestamp,omitempty"`
```
Maybe `LastUpdated` is a more clear name for this?

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
+
+	// BGPStatus holds node BGP status.
+	BGPStatus CalicoNodeBGPStatus `json:"bgpStatus,omitempty"`
```
given this is all within the `status` field anyway, we can probably shorten to just `bgp`.

e.g., 

```
status:
  bgp:
```
rather than
```
status:
  bgpStatus:
```

Not really a big deal though so if you want to keep this way I don't mind.

song-jiang replied:
Yeah, make sense.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	Peers []CalicoNodePeers `json:"peers,omitempty"`
+
+	// Routes represents routes on the node.
+	Routes []CalicoNodeRoutes `json:"routes,omitempty"`
```
I think Routes probably belongs outside of BGP?

We can learn routes from more than just BGP (e.g., from the kernel, or static routes, or from OSPF if we ever enable that).

We probably want each route object to have a field indicating from where it was learned.

song-jiang replied:
Agree. Let me move it out of BGP

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	Since string `json:"since,omitempty"`
+
+	// The reason it's in the current state.
+	Reason string `json:"info,omitempty"`
```
json tag doesn't match the field name

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	// InstalledBy indicates who installed this route.
+	// If it is populated by a BGP peer, this is the name of the BGPPeer object.
+	// If it is populated by node mesh, this is the name of the node.
+	InstalledBy string `json:"installedBy,omitempty"`
```
"LearnedFrom" rather than "installedBy"?

installed seems to suggest programming of the route locally, whereas this field actually represents where the local node learned the route from.

Also, as suggested, not all learned routes may be installed.

song-jiang commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+	Version string `json:"version,omitempty"`
+
+	// Route ID used by bird.
+	RouteID string `json:"routeID,omitempty"`
```
Yes. From bird output `Router ID is 172.20.41.231`

caseydavenport replied:
Yes, but the field is called `RouteID` when it should be `RouterID`

song-jiang commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+// CalicoNodePeers contains the status of BGP routes on the node.
+type CalicoNodeRoutes struct {
```
Yes. I haven't find a way to get a clear indication on that. I suggest to leave it for next iteration.

caseydavenport replied:
I think we need this in the first iteration, otherwise it is left unspecified and it's not clear what the user is viewing.

We can say "We only show the RIB" or "We only show the FIB", but it needs to be stated explicitly.

song-jiang commented:
@caseydavenport I've addressed your comments and also added support for dual-stack. I feel it is better to make it clear a peer/route is v4 or v6.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
```
This should have a more detailed user facing comment.

e.g., "Classes declares the types of information to monitor for this calico/node, and allows for selective status reporting about certain subsets of information."

It might be better to rename this field to something more obivous, as well.

e.g.,

```yaml
monitorStatusFor:
- Agent
- BGP
- Routes
```

or 

```yaml
reportStatusFor:
```

song-jiang replied:
I'll update the comment but I still feel `Classes` is a good name for the field.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	BGP CalicoNodeBGPStatus `json:"bgp,omitempty"`
+
+	// Route represents routes on the node.
+	Route CalicoNodeRouteStatus `json:"route,omitempty"`
```
should probably be `Routes` (plural)

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+// CalicoNodeAgentStatus defines the observed state of agent status on the node.
+type CalicoNodeAgentStatus struct {
+	// Bird4 represents the latest observed status of bird4.
+	Bird4 CalicoNodeBirdStatus `json:"bird4,omitempty"`
```
`birdv4` and `birdv6`? Everywhere else we do v4 vs v6 we include the `v`.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+// CalicoNodeRouteStatus defines the observed state of routes status on the node.
+type CalicoNodeRouteStatus struct {
+	// V4 represents IPv4 routes on the node.
+	V4 []CalicoNodeRoute `json:"bird4,omitempty"`
```
`IPv4Routes` and `IPv6Routes` - no need to abbreviate.

Also the json tag doesn't match the name of the field.

neiljerram replied:
But maybe V4Routes and V6Routes, for consistency with other use of "V4" and "V6" in this API object?

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+// CalicoNodeBirdStatus defines the observed state of bird.
+type CalicoNodeBirdStatus struct {
+	// Ready indicates if bird status is ready.
+	Ready bool `json:"ready,omitempty"`
```
API conventions say to avoid using boolean fields for things like status.

Instead, we should use an enum.

e.g.,

```yaml
status: Ready | NotReady
```

caseydavenport replied:
This comment still needs to be addressed.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	// with ListenPort set in BGPConfiguration, then we use that port to peer.
+	PeerIP string `json:"peerIP,omitempty"`
+
+	// The type is type of bgp session state.
```
I don't think "type of bgp session state" is something users will understand at a glance.

"Type indicates whether this peer is configured via the node-to-node mesh, or via en explicit global or per-node BGPPeer object"

^ I'd suggest something like this, which is more descriptive for someone reading the API docs.

neiljerram replied:
Yes, I was expecting Established, Connected etc. here.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	// The state is the bgp session state.
+	State string `json:"state,omitempty"`
+
+	// Since is the time since the condition last changed.
```
"since the state or reason last changed" ?

"condition" isn't a field in this object

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	Interface string `json:"interface,omitempty"`
+
+	// LearnedFrom indicates who installed this route.
+	// If it is populated by a BGP peer, this is the name of the BGPPeer object.
```
I think we should show the IP address of the router that sent us this route, since a BGPPeer object might configure multiple peerings with upstream routers of different IPs and so simply showing the BGPPeer name is not specific enough.

For `Kernel`, `Direct`, `Static` we should have constants and reference those.

We also probably want this field to be able to be cross-referenced with the identifying information available in the `CalicoNodePeer` object, so that it's easy to compare routes with the BGP peer that originated them (for routes learned over BGP)

That way it's easy to look at routes and find 

Might need to be structured data to represent what we need.

```yaml
learnedFrom:
  sourceType: Kernel | Direct | Static | NodeMesh | BGPPeer

  # If source is NodeMesh or BGPPeer, IP address of the router that sent us this route. 
  peer: 10.0.0.1:179

  # If source is a Kubernetes node running Calico, the name of the Kubernetes node that originated the route.
  node: k8s-node-1
```

song-jiang replied:
Yes. Make sense.

caseydavenport replied:
Comment on this is out of date now. It should probably just say "LearnedFrom contains information regarding where this route originated"

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+const (
+	NodeStatusClassTypeAgent NodeStatusClassType = "Agent"
+	NodeStatusClassTypeBGP                       = "BGP"
+	NodeStatusClassTypeRoute                     = "Route"
```
Should probably be plural, since there are many routes.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	// If port number is given, format should be `[<IPv6>]:port` or `<IPv4>:<port>` for IPv4.
+	// If optional port number is not set, and this peer IP and ASNumber belongs to a calico/node
+	// with ListenPort set in BGPConfiguration, then we use that port to peer.
+	PeerIP string `json:"peerIP,omitempty"`
```
I don't think I understand how the port comes into play on this field - is there always a `:port` set? Or is that only included if using a non-standard port?

I think we can simplify this entire thing down to something like this:

```go
// IP address and port of this BGP peer, separated by a colon. 
PeerIP string `json"peerIP,omitempty"`
```

song-jiang replied:
I copied the field from BGPPeer but you are right, in our case all the information we get is the IP address.

neiljerram replied:
Are you saying, @caseydavenport , that the port must **always** be present?  If so, fine.  If not, there can be ambiguity in some IPv6 cases if `[` `]` are not used.

caseydavenport replied:
Yeah we'd need to properly format the address if we do that. The reason to always specify the port is just to make the API more consumable - it's confusing when fields have data which is formatted conditionally based on another field, or in this case something that isn't even in this API strucutre.

We could either have separate IP / port fields, or just always include the port in this field. I am probably OK with either, although separate fields is more explicit.

caseydavenport replied:
@song-jiang what was the resolution on this one - are we just not including the port at all?

caseydavenport reviewed (changes requested):
A few more comments and questions

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+type NodeStatusClassType string
+
+const (
+	NodeStatusClassTypeAgent NodeStatusClassType = "Agent"
```
Hmm, I don't think "Agent" makes sense, unless it is assumed that CalicoNodeStatus is only for routing-related information.  Or unless it is expected that more fields will be added to CalicoNodeAgentStatus to report the liveness of Felix, confd etc.

song-jiang replied:
Naming is hard. Hmm.. I would expect information from Felix, confd etc. will be added later.

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	// Route ID used by bird.
+	RouteID string `json:"routeID,omitempty"`
+
+	// ServerTime holds the value of serverTime from birdctl output.
```
"birdctl" should be either "bird.ctl" (the socket name) or "birdcl" (the client binary name).

neiljerram replied:
and similarly below.

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	Type BGPPeerType `json:"type,omitempty"`
+
+	// The state is the bgp session state.
+	State string `json:"state,omitempty"`
```
Think this needs an RFC reference.  And maybe enumerating the possible states.

song-jiang replied:
https://github.com/projectcalico/bird/blob/feature-ipinip/proto/bgp/bgp.c#L1560 
I think this has the state information.

caseydavenport replied:
Agree with Neil on this one - this should be in terms of the RFC and not in terms of BIRD session state, since the value here is really not tied to BIRD specifically in any way.

This should probably also be an Enum type with the accepted values instead of an opaque string - e.g., `Wait`, `Connect`, `Established`, etc.,

song-jiang replied:
It seems bird session state is actually state from RFC doc https://datatracker.ietf.org/doc/html/rfc4271#section-8.2.2 
I'll add Enum type for it.

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	Since string `json:"since,omitempty"`
+
+	// The reason it's in the current state.
+	Reason string `json:"reason,omitempty"`
```
I don't understand.  What are the possible values here?

song-jiang replied:
I think probably we should just keep the field from `show protocol` output without interpreting by ourself. User can reference bird doc for the actual meaning of the field. We show exactly what bird is showing on each column. 

```
sh-4.4# birdcl -s /var/run/calico/bird.ctl show protocols 
BIRD v0.3.3+birdv1.6.8 ready.
name     proto    table    state  since       info
kernel1  Kernel   master   up     2021-08-10  
device1  Device   master   up     2021-08-10  
direct1  Direct   master   up     2021-08-10  
static1  Static   master   up     2021-08-10  
Mesh_172_20_55_6 BGP      master   up     2021-08-10  Established  
```

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
```
Would it be more conventional to use a `*metav1.Duration` field like we do in FelixConfiguration?

song-jiang replied:
This is to make sure user doesn't set the interval too short, say 10 ms.

caseydavenport replied:
I think a `metav1.Duration` is the right type for this - good thought Neil.

@song-jiang we should handle the interval with validation to make sure it's not set too small. I think even 1s might be too small of a number here, so using an `int` isn't really effective protection.

song-jiang replied:
@caseydavenport I still think 1s minimal interval makes sense. Same as linux "watch" command. If nothing changes, there is no update so 1s probably fine. I don't find a good value of minimal interval other than 1s.

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
```
I think I'd probably want a separate `NumNotEstablished` field for v4 and v6.  Then, in a typical system where I'm only interested in v4, I can implement a very simple sanity check as `v4.NumNotEstablished == 0`.

neiljerram replied:
Whereas if I see NumEstab 2 NumNot 2, it probably means missing IPv6 support, but I can't immediately be sure about that.

song-jiang replied:
Yes. Make sense.

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	V6Peers []CalicoNodePeer `json:"v6Peers,omitempty"`
+}
+
+// CalicoNodeRouteStatus defines the observed state of routes status on the node.
```
"status defines ... the state ... of status ..."  Hmm, I think we could say that better!

Also I think the problem is that we are not being specific enough.  "CalicoNodeRouteStatus" could mean "ip r", for example.  I think the name should be CalicoNodeBGPRouteStatus, and the comment "reports routes known to the Calico BGP daemon on the node"

caseydavenport replied:
The route table we're reading from BIRD isn't BGP specific though, right @neiljerram ? This is learning routes from the kernel, static, bgp protocols. If, theoretically, we add OSPF support or someone modifies a template to enable it, those routes will show up in the call we're making to birdcl as well I think.

caseydavenport replied:
Which, related to that, do we need a field indicating whether or not the route has been advertised to our BGP peers?

neiljerram commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,201 @@
+	V6 []CalicoNodeRoute `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBirdStatus defines the observed state of bird.
```
Perhaps BGPDaemonStatus, so as not to be BIRD-specific?

song-jiang replied:
The fields defined in the structure is BIRD-specific.

caseydavenport replied:
BIRD is an acronym so should be CalicoNodeBIRDStatus at least.

caseydavenport replied:
I do agree it would be better for this to be `BGPDaemonStatus` though.

I don't think that any of the fields here are actually BIRD specific - any BGP agent will have a ready state, a version, a router ID, and the listed times. Any that happen to not have one can just have that field omitted.

neiljerram reviewed (changes requested):
Some minor thoughts from me.

song-jiang commented:
> BTW - and I don't know if this is even possible - but did you consider a model where status is reported as a status subresource of the Calico Node object?

Not sure why sometimes I can't reply inline. 

@neiljerram Calico Node object is virtually kubernetes node object in KDD mode. I would expect it keeps being updated by kubelet. I feel it is better to make it a independent resource as user need to specify a spec as well.

song-jiang commented:
> Kubernetes generally uses Conditions in status to report on things like this
If we follow the rule that either the object has been updated successfully across all the fields or there is no update at all, then there is no need for `AdditionalInfo` . I feel `Condition` is used for the object who has transient states and may not suitable for NodeStatus. I'm not 100% sure what we should report back to user at all except last snapshot of the status.

caseydavenport commented:
> If we follow the rule that either the object has been updated successfully across all the fields or there is no update at all, then there is no need for AdditionalInfo . I feel Condition is used for the object who has transient states and may not suitable for NodeStatus. I'm not 100% sure what we should report back to user at all except last snapshot of the status.

Conditions are useful for asserting that the controller managing this object is healthy. In general, the lack of a condition on an object that supports them is considered to be status "Unknown". 

If we were to use conditions, we'd want something like this to indicate health.

```yaml
status:
  conditions:
    - type: "Ready"
      status: "True"
      since: "20:15:23"
```

^ that shows that the controller has been heal since the given time. We'd modify that condition in case something bad happened.

```yaml
status:
  conditions:
    - type: "Ready"
      status: "False"
      since: "20:15:27"
      reason: "BirdError"
      message: "failed to connect to BIRD socket: <error>"
```

The nice thing about `conditions` is that they are automatically handled by lots of existing tooling in the k8s ecosystem, since they are a conventional way of showing status.

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,235 @@
+	// Ready indicates if bird status is ready.
+	Ready bool `json:"ready,omitempty"`
+
+	// Bird version.
```
BIRD version

Or, ideally, just `Version of the BGP daemon`

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,235 @@
+	// Since the state or reason last changed.
+	Since string `json:"since,omitempty"`
+
+	// Extra information from bird on the bgp session.
```
I still don't really understand what sort of values will go here - could you give some examples?

I'd like to stay away from having fields that just blindly copy data from BIRD and expose it in the API

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,235 @@
+	SourceType CalicoNodeRouteSourceType `json:"sourceType,omitempty"`
+
+	// If sourceType is NodeMesh or BGPPeer, IP address of the router that sent us this route.
+	PeerIP string `json:"peerIP,omitempty"`
```
Will this be just an IP address or an IP and port? The comment suggests this is just an IP, which is probably OK.

song-jiang replied:
It is just an IP.

song-jiang replied:
Added validation of `validate:"omitempty,ip"`

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,235 @@
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds *int `json:"updateIntervalInSeconds,omitempty"`
```
Actually, to the conventions, this should be `UpdatePeriodSeconds` - https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions

caseydavenport commented on pkg/apis/projectcalico/v3/nodestatus.go:
```diff
@@ -0,0 +1,242 @@
+
+	// UpdatePeriodSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. Maximum update period is one day.
+	UpdatePeriodSeconds *uint32 `json:"updatePeriodSeconds,omitempty" validate:"required,gte=0,lte=86,400"`
```
should 86,400 just be 86400? I don't know if the validator knows how to parse that.
